platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/MergeProvider.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.provider;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.Action;
+import org.gradle.api.Task;
+import org.gradle.api.provider.Provider;
+import org.gradle.internal.Cast;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A provider that takes a list of providers of values and provides a list of those values.
+ *
+ * This effectively converts a {@code List<Provider<T>>} to a {@code Provider<List<T>>}.
+ *
+ * @param <R> The type of the values that all source providers must share.
+ */
+public class MergeProvider<R> extends AbstractMinimalProvider<List<R>> {
+
+    private final List<Provider<R>> items;
+
+    public MergeProvider(List<Provider<R>> items) {
+        this.items = items;
```

Comment:
What test are you referring to? `provider is live`? That one only modifies the providers originally passed into the constructor.

I agree users who pass in a list should not modify it after the fact. I wonder if its worth the defensive copy here given that this is an internal type.

Reply:
Yes `provider is live`.  I think it's weird that the Provider is live like this, the name `MergeProvider` doesn't suggest this behavior to me.

I think if it's worth all the immutable builders internally, then it's worth the defensive copy.

Reply:
But thats literally what a provider is supposed to do, right? The whole point is that they're lazy. The constructor should not realize the value of all providers passed into the constructor, that breaks all laziness that providers offer. 

Take a look at `FlatMapProvider` for example. That stores a reference to the provider. Otherwise it would not be lazy

Reply:
The argument to the constructor is a `List` of `Provider`s.  I'm not saying the `MergeProvider` should call `.get()` on those list elements.  I'm saying that is the consumer adds items to the list it gave the `MergeProvider` after it's constructed, that should have no affect on the `MergeProvider`.

Reply:
Oh, I see, I misread that test.  That test is fine.  A test like this should pass: 

```
def first = objects.property(String)
def second = objects.property(String)
def list = [first, second]

def provider = new MergeProvider(list)

when:
first.set("1")
second.set("2")

then:
provider.get() == ["1", "2"]

when:
def third = objects.property(String)
third.set("3")
list.add(third)

then:
provider.get() == ["1", "2"] // NOT "1", "2", "3"
```

Sorry for the confusion.

Reply:
Thats not what `provider is live` does. It does not add elements to the list passed in the constructor

Reply:
Added a defensive copy